cdap-app-fabric/src/main/java/io/cdap/cdap/security/auth/MessagingAuditLogWriter.java

```diff
@@ -0,0 +1,131 @@
+import io.cdap.cdap.api.messaging.TopicNotFoundException;
+import io.cdap.cdap.api.security.AccessException;
+import io.cdap.cdap.api.service.ServiceUnavailableException;
+import io.cdap.cdap.common.conf.CConfiguration;
+import io.cdap.cdap.common.conf.Constants;
+import io.cdap.cdap.common.service.RetryStrategies;
+import io.cdap.cdap.common.service.RetryStrategy;
+import io.cdap.cdap.messaging.DefaultTopicMetadata;
+import io.cdap.cdap.messaging.client.StoreRequestBuilder;
+import io.cdap.cdap.messaging.spi.MessagingService;
+import io.cdap.cdap.proto.id.NamespaceId;
+import io.cdap.cdap.proto.id.TopicId;
+import io.cdap.cdap.security.spi.authorization.AuditLogContext;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.Collections;
+import java.util.Queue;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * This class receives a collection of {@link AuditLogContext} and writes them in order to a
+ * messaging service / topic  ( ex - tms )
+ */
+public class MessagingAuditLogWriter implements AuditLogWriter {
+
+  private static final Logger LOG = LoggerFactory.getLogger(MessagingAuditLogWriter.class);
+  private static final Gson GSON = new Gson();
+
+  private final TopicId topic;
+  private final MessagingService messagingService;
+  private final RetryStrategy retryStrategy;
+
+  @Inject
+  public MessagingAuditLogWriter(CConfiguration cConf, MessagingService messagingService) {
+    this.topic =  NamespaceId.SYSTEM.topic(cConf.get(Constants.AuditLogging.AUDIT_LOG_EVENT_TOPIC));
+    this.messagingService = messagingService;
+    this.retryStrategy = RetryStrategies.fromConfiguration(cConf, Constants.AuditLogging.AUDIT_LOG_WRITER_RETRY_PREFIX);
+  }
+
+  /**
+   * pushes the collection of log entry to respective messaging topic
+   *
+   * @param auditLogContexts
+   */
+  @Override
+  public void publish(Queue<AuditLogContext> auditLogContexts) throws IOException {
+
+    if (auditLogContexts.isEmpty()){
+      return;
+    }
+
+    int failureCount = 0;
+    long startTime = -1L;
+    boolean done = false;
+
+    while (!done) {
+      try {
+
+        //Iterate through the queue and publish
+        while (!auditLogContexts.isEmpty()) {
+          messagingService.publish(StoreRequestBuilder.of(topic)
+                                     .addPayload(GSON.toJson(auditLogContexts.remove()))
+                                     .build());
+        }
+        done = true;
+      } catch (IOException | AccessException e) {
+        throw Throwables.propagate(e);
+      } catch (TopicNotFoundException | ServiceUnavailableException e) {
+
+        if (e instanceof TopicNotFoundException) {
+          createTopicIfNeeded();
+        }
+
+        // These exceptions are retry-able due to TMS not completely started
+        if (startTime < 0) {
+          startTime = System.currentTimeMillis();
+        }
+        long retryMillis = retryStrategy.nextRetry(++failureCount, startTime);
```

Comment:
Any reason not to use io.cdap.cdap.common.service.Retries#callWithRetries?

Reply:
I was following `MessagingProgramStatePublisher` , but yes  Retries#callWithRetries is much cleaner.
Changed it.